diff_test.go

```diff
@@ -28,151 +29,355 @@ func ExampleDefinition() AppDefinition {
-	newCfg.Services[0].Components = append(newCfg.Services[0].Components, ComponentConfig{
-		ComponentName: "service1component2",
-		InstanceConfig: InstanceConfig{
-			Image: generictypes.MustParseDockerImage("foobar"),
-		},
-	})
+		var oldDef V2AppDefinition
+		if err := json.Unmarshal([]byte(rawOldDef), &oldDef); err != nil {
+			t.Fatalf("failed to unmarshal service definition: %#v", err)
+		}
 
-	expectedDiffItems := []DiffInfo{
-		DiffInfo{Type: InfoNodeAdded, Name: []string{"app", "service1", "service1component1#changed"}},
-		DiffInfo{Type: InfoNodeAdded, Name: []string{"app", "service1", "service1component2"}},
-		DiffInfo{Type: InfoNodeRemoved, Name: []string{"app", "service1", "service1component1"}},
-	}
+		rawNewDef := `
+		{
+			"name": "redis-example-2",
+			"components": {
+				"redis1": {
+					"image": "redis",
+					"ports": 6000
+				},
+				"service1": {
+					"image": "giantswarm/redis-example:0.3.0",
+					"ports": 80,
+					"domains": { "80": "bar.com" },
+					"links": [
+						{ "component": "redis1", "target_port": 6000 }
+					]
+				},
+				"redis2": {
+					"image": "redis",
+					"ports": 6000
+				},
+				"service2": {
+					"image": "giantswarm/redis-example:0.3.0",
+					"ports": 80,
+					"domains": { "80": "foo.com" },
+					"links": [
+						{ "component": "redis2", "target_port": 6000 }
+					]
+				},
+				"redis3": {
+					"image": "redis",
+					"ports": 6379
+				},
+				"service3": {
+					"image": "giantswarm/redis-example:0.3.0",
+					"ports": 80,
+					"domains": { "80": "foo.com" },
+					"links": [
+						{ "component": "redis3", "target_port": 6379 }
+					]
+				}
+			}
+		}
+	`
+
+		var newDef V2AppDefinition
+		if err := json.Unmarshal([]byte(rawNewDef), &newDef); err != nil {
+			t.Fatalf("failed to unmarshal service definition: %#v", err)
+		}
 
-	testDiffCallWith(t, newCfg, oldCfg, expectedDiffItems)
+		expectedDiffInfos := []DiffInfo{
+			DiffInfo{
+				Type: DiffInfoServiceNameUpdated,
+				Old:  "redis-example",
+				New:  "redis-example-2",
+			},
+			DiffInfo{
+				Type: DiffInfoComponentAdded,
+				New:  "redis1",
+			},
+			DiffInfo{
+				Type: DiffInfoComponentAdded,
+				New:  "service1",
+			},
+			DiffInfo{
+				Type: DiffInfoComponentUpdated,
+				Old:  "redis2",
+				New:  "redis2",
+			},
+			DiffInfo{
+				Type: DiffInfoComponentUpdated,
+				Old:  "service2",
+				New:  "service2",
+			},
+			DiffInfo{
+				Type: DiffInfoComponentRemoved,
+				Old:  "redis",
+			},
+			DiffInfo{
+				Type: DiffInfoComponentRemoved,
+				Old:  "service",
+			},
+		}
+
+		testDiffCallWith(t, oldDef, newDef, expectedDiffInfos)
+	}
 }
 
-func TestDiffComponentMultipleChanges(t *testing.T) {
-	oldCfg := ExampleDefinition()
-	newCfg := ExampleDefinition()
+// Because of our smart data types some definitions can be represented
+// differently. The following test ensures that a diff only is created in case
+// the component really changed.
+func TestDiffComponentDefinitionNoUpdate(t *testing.T) {
```

Comment:
I don't like the name either. Did not find a better one though. It is to ensure that a component that is not updated actually will not be updated. That is, not recognized as being different.